Improve performance for distinct queries

Jackie-Jiang opened the pull request:
## Description
Currently the distinct queries are handled as aggregation, which has limitations on early termination and special type handling for better performance.

This PR adds new operators for distinct queries.
The new operators:
- Early terminate the distinct only queries when enough distinct records are collected
- Specialize the handling for different data types, using primitives if possible
- Specialize the handling for single column distinct queries
- Inner segment, the operator will calculate the distinct records using dictionary ids if possible, and only read the dictionary when returning the results

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/plan/CombinePlanNode.java:
```diff
@@ -172,7 +173,19 @@ public CombinePlanNode(List<PlanNode> planNodes, QueryContext queryContext, Exec
           _streamObserver);
     }
-    if (QueryContextUtils.isAggregationQuery(_queryContext)) {
+    if (QueryContextUtils.isSelectionQuery(_queryContext)) {
```
Common case (of aggr/group-by) should be first?

Jackie-Jiang replied:
Done

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java:
```diff
@@ -123,8 +124,13 @@ public Plan makeInstancePlan(List<IndexSegment> indexSegments, QueryContext quer
   public PlanNode makeSegmentPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
-    if (QueryContextUtils.isAggregationQuery(queryContext)) {
-      // Aggregation query
+    if (QueryContextUtils.isSelectionQuery(queryContext)) {
```
Same here, common case first?

Jackie-Jiang replied:
Done

mayankshriv commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/DistinctAggregationFunction.java:
```diff
@@ -18,31 +18,25 @@
  */
 @SuppressWarnings("rawtypes")
-public class DistinctAggregationFunction implements AggregationFunction<DistinctTable, Comparable> {
+public class DistinctAggregationFunction implements AggregationFunction<Object, Comparable> {
```
Seems this class does not implement any of the methods?

Jackie-Jiang replied:
Yes, currently `DISTINCT` keyword is not supported in the `PinotQuery`, so we use aggregation to pass the arguments. We keep it the same way for backward-compatibility, and I have added a TODO to clean it up later.

siddharthteotia replied:
DISTINCT is supported in PinotQuery as a function. The CalciteSqlParser compiles Distinct specifically as a function expression.

```
private static List<Expression> convertDistinctSelectList(SqlNodeList selectList) {
    List<Expression> selectExpr = new ArrayList<>();
    selectExpr.add(convertDistinctAndSelectListToFunctionExpression(selectList));
    return selectExpr;
  }
```

mayankshriv commented:
Some of the changes (special-casing) seem motivated by performance improvement. If so, are there any performance improvement numbers to share?

Jackie-Jiang commented:
@mayankshriv Thanks for the review. Addressed all the comments and updated the description to include the number for a simple distinct query. All the cases should perform at least the same as before (when there are multiple distinct columns and has raw index column, we fall back to the old way of using `DistinctTable` to merge raw records)

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/operator/combine/DistinctCombineOperator.java:
```diff
@@ -0,0 +1,82 @@
+  protected void mergeResultsBlocks(IntermediateResultsBlock mergedBlock, IntermediateResultsBlock blockToMerge) {
+    // TODO: Use a separate way to represent DISTINCT instead of aggregation.
+    List<Object> mergedResults = mergedBlock.getAggregationResult();
+    assert mergedResults != null && mergedResults.size() == 1 && mergedResults.get(0) instanceof DistinctTable;
```
Why not use Preconditions.checkState()? Are we moving away from them as a convention?

Jackie-Jiang replied:
assert has no performance impact on production code. We don't really need to check it, this is more for readability

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/operator/combine/DistinctCombineOperator.java:
```diff
@@ -0,0 +1,82 @@
+  @Override
+  protected boolean isQuerySatisfied(IntermediateResultsBlock resultsBlock) {
+    if (_queryContext.getOrderByExpressions() == null) {
+      List<Object> result = resultsBlock.getAggregationResult();
```
You can probably stash away the order by presence in a boolean instance variable and avoid the method call to `getOrderByExpressions()` here since I believe it will be called by each combine thread for each segment it is processing in a loop. Simply check for the instance variable

Jackie-Jiang replied:
Done

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/query/distinct/DistinctExecutorFactory.java:
```diff
@@ -0,0 +1,176 @@
+    List<ExpressionContext> expressions = distinctAggregationFunction.getInputExpressions();
+    List<OrderByExpressionContext> orderByExpressions = distinctAggregationFunction.getOrderByExpressions();
+    int limit = distinctAggregationFunction.getLimit();
+    if (orderByExpressions == null) {
```
please consider separating this into multiple private methods for better readability

Jackie-Jiang replied:
Done

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/query/distinct/dictionary/BaseDictionaryBasedMultiColumnDistinctExecutor.java:
```diff
@@ -0,0 +1,95 @@
+import org.apache.pinot.spi.utils.ByteArray;
+
+
+abstract class BaseDictionaryBasedMultiColumnDistinctExecutor implements DistinctExecutor {
```
javadoc please

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/query/distinct/dictionary/BaseDictionaryBasedSingleColumnDistinctExecutor.java:
```diff
@@ -0,0 +1,69 @@
+import org.apache.pinot.spi.utils.ByteArray;
+
+
+abstract class BaseDictionaryBasedSingleColumnDistinctExecutor implements DistinctExecutor {
```
javadoc please

siddharthteotia commented on pinot-core/src/main/java/org/apache/pinot/core/query/distinct/raw/RawMultiColumnDistinctExecutor.java:
```diff
@@ -0,0 +1,83 @@
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+
+
+public class RawMultiColumnDistinctExecutor implements DistinctExecutor {
```
Please add javadocs to all classes

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/InnerSegmentAggregationSingleValueQueriesTest.java:
```diff
@@ -175,26 +178,23 @@ public void testVeryLargeAggregationGroupBy() {
   @Test
   public void testSingleColumnDistinct() {
```
Can we add another 2 similar tests for single and multi column distinct on dictionary encoded columns. Here both the tests are on raw columns (and so we use DistinctTable)

Jackie-Jiang replied:
`DistinctQueriesTest` has all the scenarios covered

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:
```diff
@@ -170,163 +199,360 @@ private ImmutableSegment createSegment(int index, List<GenericRow> records)
   }
 
+  @Test
+  public void testSingleColumnDistinctInnerSegment() {
```
We should separate this into multiple individual unit tests.

Jackie-Jiang replied:
Split into `testSingleColumnDistinctOnlyInnerSegment` and `testSingleColumnDistinctOrderByInnerSegment`. I don't think we need to split further because that is not as easy to manage

siddharthteotia commented on pinot-core/src/test/java/org/apache/pinot/queries/DistinctQueriesTest.java:
```diff
@@ -394,285 +614,273 @@ private DistinctTable getDistinctTableInnerSegment(String query, boolean isPql)
+      String sqlQuery = sqlQueries[0];
+
+      // Check data schema
+      BrokerResponseNative pqlResponse = getBrokerResponseForPqlQuery(pqlQuery);
```
Unrelated to this PR - Can we please add another version of tests for SQL? Or may be just use `getBrokerResponseForSqlQuery()` instead of both PQL and SQL?

Jackie-Jiang replied:
Why do we need that? On the server side, PQL and SQL are handled the same way, and we can use the same test logic for both

Jackie-Jiang commented:
@siddharthteotia Addressed all the comments, please take another look

siddharthteotia reviewed (approved):
lgtm. thanks for working on this.
